File: samples/SampleCounterV5/Stores/CounterStore.cs
using Metano.TypeScript.SolidJs;
using SampleCounterV5.Models;

namespace SampleCounterV5.Stores;

public sealed class CounterStore
{
    private readonly System.Func<Counter> _get;
    private readonly ISignalSetter<Counter> _set;

    private CounterStore()
    {
        var (get, set) = Solid.CreateSignal(Counter.Zero);
        _get = get;
        _set = set;

        Solid.CreateEffect(() =>
        {
            Console.WriteLine($"Counter has changed: {State().Count}");
        });
    }

    public Counter State() => _get();

    public void Increment() => _set.Invoke(x => x.Increment());

    public void Decrement() => _set.Invoke(x => x.Decrement());

    public static CounterStore Create() => new();
}

File: targets/js/sample-counter-v5/lib/stores/counter-store.js
/** biome-ignore-all lint/complexity/noUselessConstructor: explicit shape preserved by transpiler */
import { createEffect, createSignal } from "solid-js";
import { Counter } from "#/models";
export class CounterStore {
    _counter;
    constructor() {
        this._counter = createSignal(Counter.zero);
        createEffect(() => {
            console.log(`Counter has changed: ${this.state().count}`);
        });
    }
    state() {
        return this._counter[0]();
    }
    increment() {
        this._counter[1]((x) => x.increment());
    }
    decrement() {
        this._counter[1]((x) => x.decrement());
    }
    static create() {
        return new CounterStore();
    }
}
